README.md
## oxford_gps_eth (kinetic) - 1.2.0-1

The packages in the `oxford_gps_eth` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release oxford_gps_eth --rosdistro kinetic` on `Mon, 06 Jul 2020 18:36:47 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `1.1.1-1`
- old version: `1.1.1-1`
- new version: `1.2.0-1`

Versions of tools used:

- bloom version: `0.9.7`
- catkin_pkg version: `0.4.22`
- rosdep version: `0.19.0`
- rosdistro version: `0.8.2`
- vcstools version: `0.1.42`


## oxford_gps_eth (melodic) - 1.2.0-1

The packages in the `oxford_gps_eth` repository were released into the `melodic` distro by running `/usr/bin/bloom-release oxford_gps_eth --rosdistro melodic` on `Mon, 06 Jul 2020 18:34:06 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `1.1.1-1`
- old version: `1.1.1-1`
- new version: `1.2.0-1`

Versions of tools used:

- bloom version: `0.9.7`
- catkin_pkg version: `0.4.22`
- rosdep version: `0.19.0`
- rosdistro version: `0.8.2`
- vcstools version: `0.1.42`


## oxford_gps_eth (noetic) - 1.2.0-1

The packages in the `oxford_gps_eth` repository were released into the `noetic` distro by running `/usr/bin/bloom-release oxford_gps_eth --rosdistro noetic` on `Mon, 06 Jul 2020 18:30:35 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.1.1-1`
- new version: `1.2.0-1`

Versions of tools used:

- bloom version: `0.9.7`
- catkin_pkg version: `0.4.22`
- rosdep version: `0.19.0`
- rosdistro version: `0.8.2`
- vcstools version: `0.1.42`


## oxford_gps_eth (noetic) - 1.1.1-1

The packages in the `oxford_gps_eth` repository were released into the `noetic` distro by running `/usr/bin/bloom-release oxford_gps_eth --rosdistro noetic` on `Fri, 05 Jun 2020 20:27:56 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.1.1-1`

Versions of tools used:

- bloom version: `0.9.7`
- catkin_pkg version: `0.4.20`
- rosdep version: `0.19.0`
- rosdistro version: `0.8.2`
- vcstools version: `0.1.42`


## oxford_gps_eth (melodic) - 1.1.1-1

The packages in the `oxford_gps_eth` repository were released into the `melodic` distro by running `/usr/bin/bloom-release --rosdistro melodic oxford_gps_eth` on `Thu, 31 Oct 2019 20:07:15 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `1.0.0-0`
- old version: `1.1.0-1`
- new version: `1.1.1-1`

Versions of tools used:

- bloom version: `0.9.0`
- catkin_pkg version: `0.4.14`
- rosdep version: `0.17.1`
- rosdistro version: `0.7.5`
- vcstools version: `0.1.42`


## oxford_gps_eth (kinetic) - 1.1.1-1

The packages in the `oxford_gps_eth` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release --rosdistro kinetic oxford_gps_eth` on `Thu, 31 Oct 2019 20:03:44 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `1.0.0-0`
- old version: `1.1.0-1`
- new version: `1.1.1-1`

Versions of tools used:

- bloom version: `0.9.0`
- catkin_pkg version: `0.4.14`
- rosdep version: `0.17.1`
- rosdistro version: `0.7.5`
- vcstools version: `0.1.42`


## oxford_gps_eth (melodic) - 1.1.0-1

The packages in the `oxford_gps_eth` repository were released into the `melodic` distro by running `/usr/bin/bloom-release --rosdistro melodic oxford_gps_eth --edit` on `Mon, 28 Oct 2019 21:56:49 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `1.0.0-0`
- old version: `1.0.0-0`
- new version: `1.1.0-1`

Versions of tools used:

- bloom version: `0.9.0`
- catkin_pkg version: `0.4.14`
- rosdep version: `0.17.0`
- rosdistro version: `0.7.5`
- vcstools version: `0.1.42`


## oxford_gps_eth (kinetic) - 1.1.0-1

The packages in the `oxford_gps_eth` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release --rosdistro kinetic oxford_gps_eth --edit` on `Mon, 28 Oct 2019 21:53:01 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth.git
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `1.0.0-0`
- old version: `1.0.0-0`
- new version: `1.1.0-1`

Versions of tools used:

- bloom version: `0.9.0`
- catkin_pkg version: `0.4.14`
- rosdep version: `0.17.0`
- rosdistro version: `0.7.5`
- vcstools version: `0.1.42`


## oxford_gps_eth (melodic) - 1.0.0-0

The packages in the `oxford_gps_eth` repository were released into the `melodic` distro by running `/usr/bin/bloom-release oxford_gps_eth --rosdistro melodic` on `Fri, 07 Sep 2018 18:13:10 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.6-0`
- old version: `0.0.6-0`
- new version: `1.0.0-0`

Versions of tools used:

- bloom version: `0.6.6`
- catkin_pkg version: `0.4.8`
- rosdep version: `0.12.2`
- rosdistro version: `0.6.9`
- vcstools version: `0.1.40`


## oxford_gps_eth (kinetic) - 1.0.0-0

The packages in the `oxford_gps_eth` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release oxford_gps_eth --rosdistro kinetic` on `Fri, 07 Sep 2018 18:10:05 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.6-0`
- old version: `0.0.6-0`
- new version: `1.0.0-0`

Versions of tools used:

- bloom version: `0.6.6`
- catkin_pkg version: `0.4.8`
- rosdep version: `0.12.2`
- rosdistro version: `0.6.9`
- vcstools version: `0.1.40`


## oxford_gps_eth (indigo) - 1.0.0-0

The packages in the `oxford_gps_eth` repository were released into the `indigo` distro by running `/usr/bin/bloom-release oxford_gps_eth --rosdistro indigo` on `Fri, 07 Sep 2018 18:04:02 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.6-0`
- old version: `0.0.6-0`
- new version: `1.0.0-0`

Versions of tools used:

- bloom version: `0.6.6`
- catkin_pkg version: `0.4.8`
- rosdep version: `0.12.2`
- rosdistro version: `0.6.9`
- vcstools version: `0.1.40`


## oxford_gps_eth (melodic) - 0.0.6-0

The packages in the `oxford_gps_eth` repository were released into the `melodic` distro by running `/usr/bin/bloom-release oxford_gps_eth --track melodic --rosdistro melodic` on `Mon, 26 Mar 2018 16:32:38 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `0.0.6-0`

Versions of tools used:

- bloom version: `0.6.4`
- catkin_pkg version: `0.3.8`
- rosdep version: `0.11.8`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## oxford_gps_eth (kinetic) - 0.0.6-0

The packages in the `oxford_gps_eth` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release oxford_gps_eth --track kinetic --rosdistro kinetic` on `Mon, 26 Mar 2018 16:28:22 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.5-0`
- old version: `0.0.5-0`
- new version: `0.0.6-0`

Versions of tools used:

- bloom version: `0.6.4`
- catkin_pkg version: `0.3.8`
- rosdep version: `0.11.8`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## oxford_gps_eth (indigo) - 0.0.6-0

The packages in the `oxford_gps_eth` repository were released into the `indigo` distro by running `/usr/bin/bloom-release oxford_gps_eth --track indigo --rosdistro indigo` on `Mon, 26 Mar 2018 16:24:15 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.5-0`
- old version: `0.0.5-0`
- new version: `0.0.6-0`

Versions of tools used:

- bloom version: `0.6.4`
- catkin_pkg version: `0.3.8`
- rosdep version: `0.11.8`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## oxford_gps_eth (indigo) - 0.0.5-0

The packages in the `oxford_gps_eth` repository were released into the `indigo` distro by running `/usr/bin/bloom-release oxford_gps_eth --track indigo --rosdistro indigo` on `Thu, 10 Aug 2017 17:15:22 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.4-0`
- old version: `0.0.4-0`
- new version: `0.0.5-0`

Versions of tools used:

- bloom version: `0.5.26`
- catkin_pkg version: `0.3.5`
- rosdep version: `0.11.8`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## oxford_gps_eth (kinetic) - 0.0.5-0

The packages in the `oxford_gps_eth` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release oxford_gps_eth --track kinetic --rosdistro kinetic` on `Thu, 10 Aug 2017 17:12:40 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.4-0`
- old version: `0.0.4-0`
- new version: `0.0.5-0`

Versions of tools used:

- bloom version: `0.5.26`
- catkin_pkg version: `0.3.5`
- rosdep version: `0.11.8`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## oxford_gps_eth (kinetic) - 0.0.4-0

The packages in the `oxford_gps_eth` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release oxford_gps_eth --track kinetic --rosdistro kinetic --edit` on `Tue, 12 Apr 2016 02:51:38 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `0.0.4-0`

Versions of tools used:

- bloom version: `0.5.21`
- catkin_pkg version: `0.2.10`
- rosdep version: `0.11.4`
- rosdistro version: `0.4.5`
- vcstools version: `0.1.38`


## oxford_gps_eth (indigo) - 0.0.4-0

The packages in the `oxford_gps_eth` repository were released into the `indigo` distro by running `/usr/bin/bloom-release oxford_gps_eth --track indigo --rosdistro indigo` on `Tue, 12 Apr 2016 02:46:24 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:

- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.3-0`
- old version: `0.0.3-0`
- new version: `0.0.4-0`

Versions of tools used:

- bloom version: `0.5.21`
- catkin_pkg version: `0.2.10`
- rosdep version: `0.11.4`
- rosdistro version: `0.4.5`
- vcstools version: `0.1.38`


## oxford_gps_eth (indigo) - 0.0.3-0

The packages in the `oxford_gps_eth` repository were released into the `indigo` distro by running `/usr/bin/bloom-release oxford_gps_eth --track indigo --rosdistro indigo` on `Mon, 21 Dec 2015 16:12:14 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:
- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.2-0`
- old version: `0.0.2-0`
- new version: `0.0.3-0`

Versions of tools used:
- bloom version: `0.5.20`
- catkin_pkg version: `0.2.10`
- rosdep version: `0.11.4`
- rosdistro version: `0.4.2`
- vcstools version: `0.1.38`


## oxford_gps_eth (indigo) - 0.0.2-0

The packages in the `oxford_gps_eth` repository were released into the `indigo` distro by running `/usr/bin/bloom-release oxford_gps_eth --track indigo --rosdistro indigo` on `Mon, 14 Dec 2015 15:28:06 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:
- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: https://github.com/DataspeedInc-release/oxford_gps_eth-release.git
- rosdistro version: `0.0.1-0`
- old version: `0.0.1-0`
- new version: `0.0.2-0`

Versions of tools used:
- bloom version: `0.5.20`
- catkin_pkg version: `0.2.10`
- rosdep version: `0.11.4`
- rosdistro version: `0.4.2`
- vcstools version: `0.1.38`


## oxford_gps_eth (indigo) - 0.0.1-0

The packages in the `oxford_gps_eth` repository were released into the `indigo` distro by running `/usr/bin/bloom-release --rosdistro indigo --track indigo oxford_gps_eth --edit` on `Thu, 10 Dec 2015 19:05:18 -0000`

The `oxford_gps_eth` package was released.

Version of package(s) in repository `oxford_gps_eth`:
- upstream repository: https://bitbucket.org/DataspeedInc/oxford_gps_eth
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `0.0.1-0`

Versions of tools used:
- bloom version: `0.5.20`
- catkin_pkg version: `0.2.10`
- rosdep version: `0.11.4`
- rosdistro version: `0.4.2`
- vcstools version: `0.1.38`


# oxford_gps_eth-release


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package oxford_gps_eth
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.2.0 (2020-07-06)
------------------
* Use setuptools instead of distutils for python
  http://wiki.ros.org/noetic/Migration#Setuptools_instead_of_Distutils
* Increase CMake minimum version to 3.0.2 to avoid warning about CMP0048
  http://wiki.ros.org/noetic/Migration#Increase_required_CMake_version_to_avoid_author_warning
* Add flexibility to change GGA topic and broadcast IP for NTRIP forwarding
* Fix local frame covariance calculation
* Use cached standard deviation values instead of raw packet values because those values are multiplexed into several packets
* Add quality check of standard deviation values
* Print debugging info with ROS_DEBUG()
* Change BUILD_ASSERT() to std_assert()
* Contributors: Kevin Hallenbeck, Micho Radovnikovich

1.1.1 (2019-10-31)
------------------
* Install Python script
* Add missing rospy dependency
* Contributors: Kevin Hallenbeck, Micho Radovnikovich

1.1.0 (2019-10-28)
------------------
* Add capability to forward RTCM messages from a NTRIP caster to OxTS receivers
* Contributors: Micho Radovnikovich

1.0.0 (2018-09-07)
------------------
* Use C++11 or newer
* Publish a sensor_msgs/TimeReference message with GPS time converted to UTC
* Publish a position type string more detailed than the NavSatStatus in the fix message
* Added UTM position and heading to odometry output; Odometry twist now in local frame
* Contributors: Micho Radovnikovich, Kevin Hallenbeck

0.0.6 (2018-03-26)
------------------
* Changed default listen address from broadcast to any
* Added unit tests and rostests
* Added launch file
* Contributors: Kevin Hallenbeck

0.0.5 (2017-08-10)
------------------
* Fixed velocity utm east-north-up orientation
* Properly handle unknown covariance and fields that are not present
* Updated license for year 2017
* Updated package.xml format to version 2
* Contributors: Kevin Hallenbeck, Micho Radovnikovich

0.0.4 (2016-04-11)
------------------
* Updated license year for 2016
* Contributors: Kevin Hallenbeck

0.0.3 (2015-12-21)
------------------
* Added fix for Ubuntu Saucy
* Contributors: Kevin Hallenbeck

0.0.2 (2015-12-14)
------------------
* Added fix for Ubuntu Saucy
* Contributors: Kevin Hallenbeck

0.0.1 (2015-12-10)
------------------
* Ready for public release
* Contributors: Kevin Hallenbeck


src/dispatch.h
/*********************************************************************
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2015-2016, Dataspeed Inc.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of Dataspeed Inc. nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

#ifndef _DISPATCH_H
#define _DISPATCH_H
#include <stdint.h>

#pragma pack(push, 1)
typedef union {
  uint8_t bytes[8];
  struct {
    uint32_t :32;
    uint8_t num_sats;
    uint8_t position_mode;
    uint8_t velocity_mode;
    uint8_t orientation_mode;
  } chan0;
  struct {
    uint16_t acc_position_north; // 1e-3 m
    uint16_t acc_position_east;
    uint16_t acc_position_down;
    uint8_t age;
  } chan3;
  struct {
    uint16_t acc_velocity_north; // 1e-3 m/s
    uint16_t acc_velocity_east;
    uint16_t acc_velocity_down;
    uint8_t age;
  } chan4;
  struct {
    uint16_t acc_heading; // 1e-5 rad
    uint16_t acc_pitch;
    uint16_t acc_roll;
    uint8_t age;
  } chan5;
  struct {
    int16_t base_station_age;
    int8_t base_station_id[4];
  } chan20;
  struct {
    uint16_t delay_ms;
  } chan23;
  struct {
    uint8_t heading_quality; // 0:None 1:Poor 2:RTK Float 3:RTK Integer
  } chan27;
  struct {
    int16_t heading_misalignment_angle; // 1e-4 rad
    uint16_t heading_misalignment_accuracy; // 1e-4 rad
    uint16_t :16;
    uint8_t valid;
  } chan37;
  struct {
    int16_t undulation; // 5e-3 m
    uint8_t HDOP; // 1e-1
    uint8_t PDOP; // 1e-1
  } chan48;
} Channel;
typedef struct {
  uint8_t sync;
  uint16_t time;
  int32_t accel_x :24; // 1e-5 m/s^2
  int32_t accel_y :24; // 1e-5 m/s^2
  int32_t accel_z :24; // 1e-5 m/s^2
  int32_t gyro_x :24; // 1e-4 rad/s
  int32_t gyro_y :24; // 1e-4 rad/s
  int32_t gyro_z :24; // 1e-4 rad/s
  uint8_t nav_status;
  uint8_t chksum1;
  double latitude;
  double longitude;
  float altitude;
  int32_t vel_north :24; // 1e-4 m/s
  int32_t vel_east :24; // 1e-4 m/s
  int32_t vel_down :24; // 1e-4 m/s
  int32_t heading :24; // 1e-6 rad
  int32_t pitch :24; // 1e-6 rad
  int32_t roll :24; // 1e-6 rad
  uint8_t chksum2;
  uint8_t channel;
  Channel chan;
  uint8_t chksum3;
} Packet;
enum {
  MODE_NONE = 0,
  MODE_SEARCH = 1,
  MODE_DOPLER = 2,
  MODE_SPS = 3,
  MODE_DIFFERENTIAL = 4,
  MODE_RTK_FLOAT = 5,
  MODE_RTK_INTEGER = 6,
  MODE_WAAS = 7,
  MODE_OMNISTAR_VBS = 8,
  MODE_OMNISTAR_HP = 9,
  MODE_NO_DATA = 10,
  MODE_BLANKED = 11,
  MODE_DOPLER_PP = 12,
  MODE_SPS_PP = 13,
  MODE_DIFFERENTIAL_PP = 14,
  MODE_RTK_FLOAT_PP = 15,
  MODE_RTK_INTEGER_PP = 16,
  MODE_OMNISTAR_XP = 17,
  MODE_CDGPS = 18,
  MODE_NOT_RECOGNISED = 19,
  MODE_UNKNOWN = 20,
};
#pragma pack(pop)

static bool validatePacket(const Packet *packet) {
  if (packet->sync == 0xE7) {
    const uint8_t *ptr = (uint8_t*)packet;
    uint8_t chksum = 0;
    for (unsigned int i = 1; i < sizeof(Packet) - 1; i++) {
      chksum += ptr[i];
    }
    return chksum == packet->chksum3;
  }
  return false;
}

#define BUILD_ASSERT(cond) do { (void) sizeof(char [1 - 2*!(cond)]); } while(0)
static void dispatchAssertSizes() {
  BUILD_ASSERT(8 == sizeof(Channel));
  BUILD_ASSERT(72 == sizeof(Packet));
}
#undef BUILD_ASSERT

#endif // _DISPATCH_H
